### PR TITLE
[pattgen, dv] Transition to V1 then V2

### PR DESCRIPTION
**Pull request by tunghoang290780**
Ready for V1 and V2 review

Signed-off-by: Tung Hoang <hoang.tung@wdc.com>

### CODE REVIEW

File: hw/ip/pattgen/doc/checklist.md [hunk 1]
```diff
@@ -163,24 +163,25 @@ Review        | [V2_CHECKLIST_SCOPED][]               | Done        |
...
 Testbench     | [SIM_TB_ENV_COMPLETED][]                | Done        |
 Tests         | [SIM_ALL_TESTS_PASSING][]               | Done        |
-Tests         | [FPV_ALL_ASSERTIONS_WRITTEN][]          | N/A         |
+Tests         | [FPV_ALL_ASSERTIONS_WRITTEN][]          | Done        |
 Tests         | [FPV_ALL_ASSUMPTIONS_REVIEWED][]        | N/A         |
 Tests         | [SIM_FW_SIMULATED][]                    | N/A         |
 Regression    | [SIM_NIGHTLY_REGRESSION_V2][]           | Done        |
-Coverage      | [SIM_CODE_COVERAGE_V2][]                | Not Started |
-Coverage      | [SIM_FUNCTIONAL_COVERAGE_V2][]          | Not Started |
+Coverage      | [SIM_CODE_COVERAGE_V2][]                | N/A         |
```
Comment (cindychip) [inline hw/ip/pattgen/doc/checklist.md:176]: Is there a reason why this is N/A?

**Review by weicaiyang — COMMENTED** [review]
we need `DV_DOC_TESTPLAN_REVIEWED` to be done before we sign off V2
Do we need to review meeting for this item? @sriyerg

**Comment by cindychip** [comment]
Good point, and the new DV plan needs to have lists of coverpoints.

  Cindy Chen

**Comment by weicaiyang** [comment]
> Good point, and the new DV plan needs to have lists of coverpoints. Cindy Chen

yeah, that's the new requirement. perhaps @tunghoang290780 can use otp_ctrl as example to list all the covergroups in DV plan

File: hw/ip/pattgen/doc/checklist.md (hunk 1 above)
Comment (tunghoang290780) [reply hw/ip/pattgen/doc/checklist.md:176]: Fixed!

File: hw/ip/pattgen/doc/checklist.md [hunk 2]
```diff
@@ -163,24 +163,25 @@ Review        | [V2_CHECKLIST_SCOPED][]               | Done        |
...
+Coverage      | [SIM_FUNCTIONAL_COVERAGE_V2][]          | Done        |
 Coverage      | [FPV_CODE_COVERAGE_V2][]                | N/A         |
 Coverage      | [FPV_COI_COVERAGE_V2][]                 | N/A         |
 Code Quality  | [TB_LINT_PASS][]                        | Done        |
 Integration   | [PRE_VERIFIED_SUB_MODULES_V2][]         | N/A         |
 Issues        | [NO_HIGH_PRIORITY_ISSUES_PENDING][]     | Done        |
 Issues        | [ALL_LOW_PRIORITY_ISSUES_ROOT_CAUSED][] | Done        |
+Integration   | [PRE_VERIFIED_SUB_MODULES_V2][]         | N/A         |
```
[the remark this answers is not in the document]
Comment (tunghoang290780) [reply hw/ip/pattgen/doc/checklist.md:184]: Fixed

**Comment by tunghoang290780** [comment]
> > Good point, and the new DV plan needs to have lists of coverpoints. Cindy Chen
> 
> yeah, that's the new requirement. perhaps @tunghoang290780 can use otp_ctrl as example to list all the covergroups in DV plan

@weicaiyang I've updated the DV plan to include covergroups

File: hw/ip/pattgen/data/pattgen_testplan.hjson [hunk 3]
```diff
@@ -81,4 +81,16 @@
...
+  covergroups: [
+    {
+      name: pattgen_op_cg
+      desc: '''
+            Covers the following conditions:
+              - whether single or dual channel are enabled
+              - whether channel polarity bit is set/unset
+              - whether channel are run at high/low throughput
+            '''
+    }
```
Comment (weicaiyang) [inline hw/ip/pattgen/data/pattgen_testplan.hjson:94]: we can consider to cover every bit of data and predivider are toggled. use auto bins can do it easily. there are some examples [here](https://verificationguide.com/systemverilog/systemverilog-functional-coverage/)

and also cover some value of data length and number of repetitions.

Looks like we should have a review meeting to go over these.

**Comment by sriyerg** [comment]
The following items are gating the transition to V2:

- [ ] All V1 and V2 tests passing: https://reports.opentitan.org/hw/ip/pattgen/dv/latest/results.html
- [ ] Pattgen coverage not getting generated in the nightly reports - @cindychip 
- [ ] Testplan (and coverage plan review) at V2 (we will deep-dive on coverage configuration & exclusions as well)

File: hw/ip/pattgen/doc/checklist.md [hunk 4]
```diff
@@ -163,18 +163,18 @@ Review        | [V2_CHECKLIST_SCOPED][]               | Done        |
  Type         | Item                                    | Resolution  | Note/Collaterals
 --------------|-----------------------------------------|-------------|------------------
 Documentation | [DESIGN_DELTAS_CAPTURED_V2][]           | N/A         |
-Documentation | [DV_DOC_COMPLETED][]                    | Not Started |
+Documentation | [DV_DOC_COMPLETED][]                    | Done        |
 Testbench     | [FUNCTIONAL_COVERAGE_IMPLEMENTED][]     | Not Started |
 Testbench     | [ALL_INTERFACES_EXERCISED][]            | Done        |
 Testbench     | [ALL_ASSERTION_CHECKS_ADDED][]          | Done        |
 Testbench     | [SIM_TB_ENV_COMPLETED][]                | Done        |
 Tests         | [SIM_ALL_TESTS_PASSING][]               | Done        |
```
Comment (sriyerg) [inline hw/ip/pattgen/doc/checklist.md:171]: PLease see: https://reports.opentitan.org/hw/ip/pattgen/dv/latest/results.html
There are several failing tests. Please triage and get all V1 and V2 tests passing before marking this as done.

File: hw/ip/pattgen/data/pattgen_testplan.hjson (hunk 3 above)
Comment (sriyerg) [reply hw/ip/pattgen/data/pattgen_testplan.hjson:94]: +1 
There are several things about pattgen that can be captured in functional coverage plan. Please work with the designer to come up with a detailed plan. We should have ideally documented it before V1.

File: hw/ip/pattgen/doc/checklist.md [hunk 5]
```diff
@@ -163,18 +163,18 @@ Review        | [V2_CHECKLIST_SCOPED][]               | Done        |
  Type         | Item                                    | Resolution  | Note/Collaterals
 --------------|-----------------------------------------|-------------|------------------
 Documentation | [DESIGN_DELTAS_CAPTURED_V2][]           | N/A         |
-Documentation | [DV_DOC_COMPLETED][]                    | Not Started |
+Documentation | [DV_DOC_COMPLETED][]                    | Done        |
```
Comment (sriyerg) [inline hw/ip/pattgen/doc/checklist.md:166]: this item is not complete either - the doc still has contains template text.

**Comment by cindychip** [comment]
Gating item 2). Last Friday we have some hanging and license issue, it should be fixed by today's regression result. 
Weekend regression was killed due to workstation random restart. 
Sorry for the inconvenience!

File: hw/ip/pattgen/doc/checklist.md (hunk 4 above)
Comment (tunghoang290780) [reply hw/ip/pattgen/doc/checklist.md:171]: @sriyerg it has been fixed by Cindy. In addition, the alert test was added in PR #7015

File: hw/ip/pattgen/data/pattgen.prj.hjson [hunk 6]
```diff
@@ -12,7 +12,7 @@
         version:            "1.0",
         life_stage:         "L1",
         design_stage:       "D1",
-        verification_stage: "V1",
+        verification_stage: "V2",
```
Comment (sriyerg) [inline hw/ip/pattgen/data/pattgen.prj.hjson:15]: Please move this to a separate PR once all checklist items are marked done.

File: hw/ip/pattgen/doc/checklist.md (hunk 5 above)
Comment (tunghoang290780) [reply hw/ip/pattgen/doc/checklist.md:166]: Fix it

File: hw/ip/pattgen/data/pattgen.prj.hjson (hunk 6 above)
Comment (tunghoang290780) [reply hw/ip/pattgen/data/pattgen.prj.hjson:15]: Roll back V1 for review

**Comment by tunghoang290780** [comment]
@sriyerg @weicaiyang @cindychip I addressed some comments and Cindy cleaned the gating items 1,2. Do you think it is enough to have V1 review then layout the missing points for V2

**Comment by weicaiyang** [comment]
> @sriyerg @weicaiyang @cindychip I addressed some comments and Cindy cleaned the gating items 1,2. Do you think it is enough to have V1 review then layout the missing points for V2

Looks like we have done more than V1 for this block, we should have a review meeting to make it to V1

**Comment by tunghoang290780** [comment]
> > @sriyerg @weicaiyang @cindychip I addressed some comments and Cindy cleaned the gating items 1,2. Do you think it is enough to have V1 review then layout the missing points for V2
> 
> Looks like we have done more than V1 for this block, we should have a review meeting to make it to V1

@weicaiyang indeed, except more functional coverage plans could be added

**Comment by tunghoang290780** [comment]
@weicaiyang @sriyerg @rasmus-madsen Could this PR enough for a merge then go to V1 review? The other remained issues can be resolved in V2.

File: hw/ip/pattgen/doc/checklist.md [hunk 7]
```diff
@@ -163,18 +163,18 @@ Review        | [V2_CHECKLIST_SCOPED][]               | Done        |
...
 Tests         | [SIM_ALL_TESTS_PASSING][]               | Done        |
-Tests         | [FPV_ALL_ASSERTIONS_WRITTEN][]          | N/A         |
+Tests         | [FPV_ALL_ASSERTIONS_WRITTEN][]          | Done        |
 Tests         | [FPV_ALL_ASSUMPTIONS_REVIEWED][]        | N/A         |
 Tests         | [SIM_FW_SIMULATED][]                    | N/A         |
 Regression    | [SIM_NIGHTLY_REGRESSION_V2][]           | Done        |
-Coverage      | [SIM_CODE_COVERAGE_V2][]                | Not Started |
-Coverage      | [SIM_FUNCTIONAL_COVERAGE_V2][]          | Not Started |
+Coverage      | [SIM_CODE_COVERAGE_V2][]                | Done        |
+Coverage      | [SIM_FUNCTIONAL_COVERAGE_V2][]          | Done        |
```
Comment (weicaiyang) [inline hw/ip/pattgen/doc/checklist.md:177]: these 2 items seem not fully done yet.

**Review by weicaiyang — COMMENTED** [review]
@tunghoang290780 we can have a testplan review meeting to claim this block to V1.

**Comment by rasmus-madsen** [comment]
Closing this PR as Alex is addressing V1 in #7816 
and will create a new PR for the V2 part
